pkg/vdri/trustbloc/didconfiguration/helpers.go

```diff
@@ -0,0 +1,212 @@
+	claims := models.DomainLinkageAssertionClaims{
+		ISS:    didValue,
+		Domain: domain,
+		Exp:    expiryTime,
+	}
+
+	claimsBytes, err := json.Marshal(claims)
+	if err != nil {
+		return nil, fmt.Errorf("can't marshal claims: %w", err)
+	}
+
+	signer, err := jose.NewSigner(*signingKey, nil)
+	if err != nil {
+		return nil, fmt.Errorf("can't construct signer: %w", err)
+	}
+
+	jws, err := signer.Sign(claimsBytes)
+	if err != nil {
+		return nil, fmt.Errorf("can't sign claims: %w", err)
+	}
+
+	jwsCompact, err := jws.CompactSerialize()
+	if err != nil {
+		return nil, fmt.Errorf("can't serialize signature: %w", err)
+	}
+
+	return &models.DomainLinkageAssertion{
+		DID: didValue,
+		JWT: jwsCompact,
+	}, nil
+}
+
+// VerifyDIDConfiguration verifies a DID configuration, using the given VDRI to resolve the DID.
+//   returns a list of the DIDs that were successfully authenticated to this domain
+func VerifyDIDConfiguration(domain string, configuration *models.DIDConfiguration, doc *did.Doc) ([]string, error) {
+	DIDSet := map[string]struct{}{}
+
+	var errs []string
+
+	for i, dla := range configuration.Entries {
+		err := ValidateDomainLinkageAssertion(domain, dla, doc)
+		if err != nil {
+			log.Debugf("domain linkage assertion %v for %s invalid", i, domain)
+
+			errs = append(errs, err.Error())
+		} else {
+			DIDSet[dla.DID] = struct{}{}
+		}
+	}
+
+	DIDs := make([]string, 0, len(DIDSet))
+	for id := range DIDSet {
+		DIDs = append(DIDs, id)
+	}
+
+	if len(DIDs) == 0 {
+		errMsg := ""
+		for _, ems := range errs {
+			errMsg += "`" + ems + "`, "
+		}
+
+		return nil, fmt.Errorf("all domain linkage assertions invalid for domain %s: %s", domain, errMsg)
+	}
+
+	return DIDs, nil
+}
+
+// ValidateDomainLinkageAssertion validates a domain linkage assertion, using the given VDRI to resolve the DID
+func ValidateDomainLinkageAssertion(domain string, assertion models.DomainLinkageAssertion, doc *did.Doc) error {
+	jws, err := jose.ParseSigned(assertion.JWT)
+	if err != nil {
+		return fmt.Errorf("cannot parse assertion JWT: %w", err)
+	}
+
+	var claims models.DomainLinkageAssertionClaims
+
+	rawClaims := jws.UnsafePayloadWithoutVerification()
+	err = json.Unmarshal(rawClaims, &claims)
+
+	if err != nil {
+		return fmt.Errorf("cannot parse assertion JWT claims: %w", err)
+	}
+
+	if claims.ISS != assertion.DID {
+		return fmt.Errorf("assertion DID does not match signature")
+	}
+
+	if claims.Domain != domain {
+		return fmt.Errorf("assertion domain does not match host domain")
+	}
+
+	if claims.Exp != 0 && claims.Exp <= time.Now().Unix() {
+		return fmt.Errorf("assertion has expired")
+	}
+
+	_, err = VerifyDIDSignature(jws, doc)
+
+	return err
+}
+
+// VerifyDIDSignature verify a signature using a DID doc
+func VerifyDIDSignature(jws *jose.JSONWebSignature, doc *did.Doc) ([]byte, error) {
+	if jws == nil {
+		return nil, fmt.Errorf("jws is nil")
+	}
+
+	verified := false
```

Comment:
Its a personal preference but its good to initialize variable right before you use for better readability.